LFS171x/fabric-material/chaincode/tuna-app/tuna-chaincode.go

```diff
@@ -0,0 +1,231 @@
+		Tuna{Vessel: "129R", Location: "153.0054, 12.6429", Timestamp: "1485153091", Holder: "Carlo"},
+		Tuna{Vessel: "49W4", Location: "51.9435, 8.2735", Timestamp: "1487745091", Holder: "Fatima"},
+	}
+
+	i := 0
+	for i < len(tuna) {
+		fmt.Println("i is ", i)
+		tunaAsBytes, _ := json.Marshal(tuna[i])
+		APIstub.PutState(strconv.Itoa(i+1), tunaAsBytes)
+		fmt.Println("Added", tuna[i])
+		i = i + 1
+	}
+
+	return shim.Success(nil)
+}
+
+/*
+ * The recordTuna method *
+Fisherman like Sarah would use to record each of her tuna catches. 
+This method takes in five arguments (attributes to be saved in the ledger). 
+ */
+func (s *SmartContract) recordTuna(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
+
+	if len(args) != 5 {
+		return shim.Error("Incorrect number of arguments. Expecting 5")
+	}
+
+	var tuna = Tuna{ Vessel: args[1], Location: args[2], Timestamp: args[3], Holder: args[4] }
+
+	tunaAsBytes, _ := json.Marshal(tuna)
+	err := APIstub.PutState(args[0], tunaAsBytes)
+	if err != nil {
+		return shim.Error(fmt.Sprintf("Failed to record tuna catch: %s", args[0]))
+	}
+
+	return shim.Success(nil)
+}
+
+/*
+ * The queryAllTuna method *
+allows for assessing all the records added to the ledger(all tuna catches)
+This method does not take any arguments. Returns JSON string containing results. 
+ */
+func (s *SmartContract) queryAllTuna(APIstub shim.ChaincodeStubInterface) sc.Response {
+
+	startKey := "0"
+	endKey := "999"
+
+	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+	defer resultsIterator.Close()
+
+	// buffer is a JSON array containing QueryResults
+	var buffer bytes.Buffer
+	buffer.WriteString("[")
+
+	bArrayMemberAlreadyWritten := false
+	for resultsIterator.HasNext() {
+		queryResponse, err := resultsIterator.Next()
+		if err != nil {
+			return shim.Error(err.Error())
+		}
+		// Add comma before array members,suppress it for the first array member
+		if bArrayMemberAlreadyWritten == true {
+			buffer.WriteString(",")
+		}
+		buffer.WriteString("{\"Key\":")
+		buffer.WriteString("\"")
+		buffer.WriteString(queryResponse.Key)
+		buffer.WriteString("\"")
+
+		buffer.WriteString(", \"Record\":")
+		// Record is a JSON object, so we write as-is
+		buffer.WriteString(string(queryResponse.Value))
+		buffer.WriteString("}")
+		bArrayMemberAlreadyWritten = true
+	}
+	buffer.WriteString("]")
+
+	fmt.Printf("- queryAllTuna:\n%s\n", buffer.String())
+
+	return shim.Success(buffer.Bytes())
+}
+
+/*
+ * The changeTunaHolder method *
+The data in the world state can be updated with who has possession. 
+This function takes in 2 arguments, tuna id and new holder name. 
+ */
+func (s *SmartContract) changeTunaHolder(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
+
+	if len(args) != 2 {
+		return shim.Error("Incorrect number of arguments. Expecting 2")
+	}
+
+	tunaAsBytes, _ := APIstub.GetState(args[0])
```

Comment:
We should check if the Tuna exists before we attempt to change the holder.

Reply:
should be done now with new commit